File: packages/format/src/h264/codec.ts
import { FormatError } from "../errors.js";

export type H264LevelIdc =
  | 10 | 11 | 12 | 13
  | 20 | 21 | 22
  | 30 | 31 | 32
  | 40 | 41 | 42
  | 50 | 51 | 52
  | 60 | 61 | 62;

export type H264Codec =
  | "avc1.64000A" | "avc1.64000B" | "avc1.64000C" | "avc1.64000D"
  | "avc1.640014" | "avc1.640015" | "avc1.640016"
  | "avc1.64001E" | "avc1.64001F" | "avc1.640020"
  | "avc1.640028" | "avc1.640029" | "avc1.64002A"
  | "avc1.640032" | "avc1.640033" | "avc1.640034"
  | "avc1.64003C" | "avc1.64003D" | "avc1.64003E";

export interface H264LevelLimits {
  readonly levelIdc: H264LevelIdc;
  readonly codec: H264Codec;
  readonly maximumMacroblocksPerSecond: number;
  readonly maximumMacroblocksPerFrame: number;
  readonly maximumMacroblockDimension: number;
  readonly maximumDpbMacroblocks: number;
  readonly maximumBitrate: number;
  readonly maximumCpbBits: number;
}

const LEVEL_ROWS = Object.freeze([
  [10, "avc1.64000A", 1_485, 99, 396, 64_000, 175_000],
  [11, "avc1.64000B", 3_000, 396, 900, 192_000, 500_000],
  [12, "avc1.64000C", 6_000, 396, 2_376, 384_000, 1_000_000],
  [13, "avc1.64000D", 11_880, 396, 2_376, 768_000, 2_000_000],
  [20, "avc1.640014", 11_880, 396, 2_376, 2_000_000, 2_000_000],
  [21, "avc1.640015", 19_800, 792, 4_752, 4_000_000, 4_000_000],
  [22, "avc1.640016", 20_250, 1_620, 8_100, 4_000_000, 4_000_000],
  [30, "avc1.64001E", 40_500, 1_620, 8_100, 10_000_000, 10_000_000],
  [31, "avc1.64001F", 108_000, 3_600, 18_000, 14_000_000, 14_000_000],
  [32, "avc1.640020", 216_000, 5_120, 20_480, 20_000_000, 20_000_000],
  [40, "avc1.640028", 245_760, 8_192, 32_768, 20_000_000, 25_000_000],
  [41, "avc1.640029", 245_760, 8_192, 32_768, 50_000_000, 62_500_000],
  [42, "avc1.64002A", 522_240, 8_704, 34_816, 50_000_000, 62_500_000],
  [50, "avc1.640032", 589_824, 22_080, 110_400, 135_000_000, 135_000_000],
  [51, "avc1.640033", 983_040, 36_864, 184_320, 240_000_000, 240_000_000],
  [52, "avc1.640034", 2_073_600, 36_864, 184_320, 240_000_000, 240_000_000],
  [60, "avc1.64003C", 4_177_920, 139_264, 696_320, 240_000_000, 240_000_000],
  [61, "avc1.64003D", 8_355_840, 139_264, 696_320, 480_000_000, 480_000_000],
  [62, "avc1.64003E", 16_711_680, 139_264, 696_320, 800_000_000, 800_000_000]
] as const);

const LEVELS = new Map<number, H264LevelLimits>(LEVEL_ROWS.map((row) => [
  row[0],
  Object.freeze({
    levelIdc: row[0],
    codec: row[1],
    maximumMacroblocksPerSecond: row[2],
    maximumMacroblocksPerFrame: row[3],
    maximumMacroblockDimension: Math.floor(Math.sqrt(row[3] * 8)),
    maximumDpbMacroblocks: row[4],
    maximumBitrate: row[5],
    maximumCpbBits: row[6]
  })
]));

const CODECS = new Map<string, H264LevelLimits>(
  [...LEVELS.values()].map((limits) => [limits.codec, limits])
);

export function isH264LevelIdc(value: number): value is H264LevelIdc {
  return LEVELS.has(value);
}

export function h264LevelLimits(levelIdc: number): Readonly<H264LevelLimits> {
  const limits = LEVELS.get(levelIdc);
  if (limits === undefined) {
    throw new FormatError("PROFILE_INVALID", "H264 level_idc is unsupported");
  }
  return limits;
}

export function h264CodecForLevel(levelIdc: number): H264Codec {
  return h264LevelLimits(levelIdc).codec;
}

export function parseH264Codec(codec: unknown): Readonly<H264LevelLimits> {
  const limits = typeof codec === "string" ? CODECS.get(codec) : undefined;
  if (limits === undefined) {
    throw new FormatError("PROFILE_INVALID", "H264 codec must identify a supported High-profile level");
  }
  return limits;
}

export function isH264Codec(codec: unknown): codec is H264Codec {
  return typeof codec === "string" && CODECS.has(codec);
}

File: flutter/packages/aval_format/lib/src/h264/codec.dart
/// High-profile H264 level table and codec-string identification.
///
/// Dart port of `packages/format/src/h264/codec.ts`.
library;

import 'dart:math' as math;

import '../errors.dart';

/// One `level_idc`, matching the closed set carried by [_levelRows].
///
/// TS `H264LevelIdc` is a string-literal union of numeric `level_idc` values;
/// used only for equality, so it is a Dart `int` typedef per the port
/// conventions.
typedef H264LevelIdc = int;

/// Codec identifier string, e.g. `"avc1.640020"` (High profile).
///
/// TS `H264Codec` is a string-literal union used only for equality/membership.
typedef H264Codec = String;

class H264LevelLimits {
  const H264LevelLimits({
    required this.levelIdc,
    required this.codec,
    required this.maximumMacroblocksPerSecond,
    required this.maximumMacroblocksPerFrame,
    required this.maximumMacroblockDimension,
    required this.maximumDpbMacroblocks,
    required this.maximumBitrate,
    required this.maximumCpbBits,
  });

  final H264LevelIdc levelIdc;
  final H264Codec codec;
  final int maximumMacroblocksPerSecond;
  final int maximumMacroblocksPerFrame;
  final int maximumMacroblockDimension;
  final int maximumDpbMacroblocks;
  final int maximumBitrate;
  final int maximumCpbBits;
}

// Columns: level_idc, codec, maxMb/s, maxMb/frame, maxDpbMb, maxBitrate,
// maxCpbBits. Mirrors `codec.ts` `LEVEL_ROWS`.
const List<List<Object>> _levelRows = [
  [10, 'avc1.64000A', 1485, 99, 396, 64000, 175000],
  [11, 'avc1.64000B', 3000, 396, 900, 192000, 500000],
  [12, 'avc1.64000C', 6000, 396, 2376, 384000, 1000000],
  [13, 'avc1.64000D', 11880, 396, 2376, 768000, 2000000],
  [20, 'avc1.640014', 11880, 396, 2376, 2000000, 2000000],
  [21, 'avc1.640015', 19800, 792, 4752, 4000000, 4000000],
  [22, 'avc1.640016', 20250, 1620, 8100, 4000000, 4000000],
  [30, 'avc1.64001E', 40500, 1620, 8100, 10000000, 10000000],
  [31, 'avc1.64001F', 108000, 3600, 18000, 14000000, 14000000],
  [32, 'avc1.640020', 216000, 5120, 20480, 20000000, 20000000],
  [40, 'avc1.640028', 245760, 8192, 32768, 20000000, 25000000],
  [41, 'avc1.640029', 245760, 8192, 32768, 50000000, 62500000],
  [42, 'avc1.64002A', 522240, 8704, 34816, 50000000, 62500000],
  [50, 'avc1.640032', 589824, 22080, 110400, 135000000, 135000000],
  [51, 'avc1.640033', 983040, 36864, 184320, 240000000, 240000000],
  [52, 'avc1.640034', 2073600, 36864, 184320, 240000000, 240000000],
  [60, 'avc1.64003C', 4177920, 139264, 696320, 240000000, 240000000],
  [61, 'avc1.64003D', 8355840, 139264, 696320, 480000000, 480000000],
  [62, 'avc1.64003E', 16711680, 139264, 696320, 800000000, 800000000],
];

// `Math.floor(Math.sqrt(row[3] * 8))` from `codec.ts:59`.
int _maximumMacroblockDimension(int macroblocksPerFrame) =>
    math.sqrt(macroblocksPerFrame * 8).floor();

final Map<int, H264LevelLimits> _levels = {
  for (final row in _levelRows)
    (row[0] as int): H264LevelLimits(
      levelIdc: row[0] as int,
      codec: row[1] as String,
      maximumMacroblocksPerSecond: row[2] as int,
      maximumMacroblocksPerFrame: row[3] as int,
      maximumMacroblockDimension: _maximumMacroblockDimension(row[3] as int),
      maximumDpbMacroblocks: row[4] as int,
      maximumBitrate: row[5] as int,
      maximumCpbBits: row[6] as int,
    ),
};

final Map<String, H264LevelLimits> _codecs = {
  for (final limits in _levels.values) limits.codec: limits,
};

bool isH264LevelIdc(int value) => _levels.containsKey(value);

H264LevelLimits h264LevelLimits(int levelIdc) {
  final limits = _levels[levelIdc];
  if (limits == null) {
    throw FormatError(
      FormatErrorCode.profileInvalid,
      'H264 level_idc is unsupported',
    );
  }
  return limits;
}

H264Codec h264CodecForLevel(int levelIdc) => h264LevelLimits(levelIdc).codec;

H264LevelLimits parseH264Codec(Object? codec) {
  final limits = codec is String ? _codecs[codec] : null;
  if (limits == null) {
    throw FormatError(
      FormatErrorCode.profileInvalid,
      'H264 codec must identify a supported High-profile level',
    );
  }
  return limits;
}

bool isH264Codec(Object? codec) =>
    codec is String && _codecs.containsKey(codec);
